File: interviewbit/sorted-permutation-rank-with-repeats.cpp
#include <bits/stdc++.h>
using namespace std;

int factorial(int n) {
    if (n == 1) return 1;
    return n % 1000003 * factorial(n - 1) % 1000003;
}
int getIndexOf(char c) {
    if ((int)c <= 90) {
        return (int)c - 65;
    }
    return (int)c - 97 + 26;
}
char getCharAtIndex(int i) {
    if (i <= 25) {
        return (char)(i + 65);
    }
    return (char)(i + 97 - 26);
}
int countPerms(string sub, string sup) {
    // counting occurences
    int occurence[52];
    memset(occurence, 0, sizeof(occurence));
    for (int i = 0; i < sup.length(); i++){
        occurence[getIndexOf(sup[i])]++;
    }
    for (int i = 0; i <= sub.length(); i++) {
        occurence[getIndexOf(sub[i])]--;
    }

    // finding repetitions
    vector<int> repetitions;
    for (int i = 0; i < 52; i++) {
        if (occurence[i] > 1) {
            repetitions.push_back(occurence[i]);
        }
    }

    // calculating permutations
    int perms;
    perms = factorial(sup.length() - sub.length());
    for (int i = 0; i < repetitions.size(); i++) {
        perms /= factorial(repetitions[i]);
    }
    return perms;
}

int main () {

    string A;
    cin >> A;

    int sol = 0, pos = 0;
    string fixed;

    int occurence[52];
    memset(occurence, 0, sizeof(occurence));
    for (int i = 0; i < A.length(); i++) {
        occurence[getIndexOf(A[i])]++;
    }

    while (pos < A.length() - 1) {
        for (int i = 0 ; i < 52; i++) {
            if (i < getIndexOf(A[pos])) {
                if (occurence[i] != 0) {
                    cout << "fixed string " << fixed + getCharAtIndex(i) << ", " << countPerms(fixed + getCharAtIndex(i), A) << " permutations added " << endl;
                    sol = (sol + countPerms(fixed + getCharAtIndex(i), A)) % 1000003;
                }
            }
            else {
                break;
            }
        }
        fixed += A[pos];
        occurence[getIndexOf(A[pos])]--;
        pos++;
    }

    cout << (sol + 1) % 1000003 << endl;

    return 0;
}

File: interviewbit/sorted_permutation_rank_with_repeats.py
class Solution:
    # @param A : string
    # @return an integer
    def findRank(self, A):
        sol = 0
        pos = 0
        fixed = ''
        
        occurence = [0 for i in range(0, 52)]
        for i in A:
            occurence[self.getIndexOf(i)] += 1
        
        while pos < len(A) - 1:
            for i in range(0, 52):
                if i < self.getIndexOf(A[pos]):
                    if occurence[i]:
                        sol = (sol + self.countPerms(fixed + self.getCharAtIndex(i), A)) % 1000003
                else:
                    break
            fixed += A[pos]
            occurence[self.getIndexOf(A[pos])] -= 1
            pos += 1
        
        return (sol + 1) % 1000003
    
    def factorial(self, n):
        if n == 0:
            return 1
        return n * self.factorial(n - 1)
    
    def getIndexOf(self, c):
        if ord(c) <= 90:
            return ord(c) - 65
        return ord(c) - 97 + 26
    
    def getCharAtIndex(self, i):
        if i <= 25:
            return chr(i + 65)
        return chr(i + 97 - 26);
    
    def countPerms(self, sub, sup):
        occurence = [0 for i in range(0, 52)]
        for i in sup:
            occurence[self.getIndexOf(i)] += 1
        for i in sub:
            occurence[self.getIndexOf(i)] -= 1
        
        repetitions = [i for i in occurence if i > 1]
        
        numerator = self.factorial(len(sup)-len(sub))
        denominator = 1
        for i in repetitions:
            denominator *= i
        return (numerator / denominator) % 1000003

